### PR TITLE
new PR requirement and commit message proposal

### PR DESCRIPTION
**Pull request by zhangyangyu**
Issue Number: ref #161 

@Mini256

### CODE REVIEW

File: votes/0162-new-pr-requirement-and-commit-message-style.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    * fix something 1
+    * fix something 2
+    ``` 
+
+The content in the `commit-message` code block will be extracted and pasted into the final commit message body. The DCO requirement is still needed. But along with the `Co-authored-by` information, they will be grouped and distinguished by the bot and be appended to the final commit message. So after the proposal, for the first sample commit message in the Motivation section, it will be like:
+
+```
+rsmeter: add pubsub datasink  (#11719)
+
+Signed-off-by: Zhenchi <zhongzc_arch@outlook.com>
```
Comment (sunxiaoguang) [inline votes/0162-new-pr-requirement-and-commit-message-style.md:136]: Can bot extract `Signed-off-by` and `Co-authored-by` from original commit message automatically without asking  author to explicitly put them into code block?

Bot should be able to keep original order of sign off messages and eliminate duplicated sign offs automatically. This way authors are only required to write down whatever messages they really care about in code block.

Comment (zhangyangyu) [reply votes/0162-new-pr-requirement-and-commit-message-style.md:136]: Yes, that's our goal. The bot will extract the `Signed-off-by` and `Co-authored-by` from the original commit messages and eliminate duplicates. The new code block should only contains any other meaningful descriptions explicitly written by the author. And I am not sure the order of sign off messages matters, but definitely not a hard work.

Comment (Mini256) [reply votes/0162-new-pr-requirement-and-commit-message-style.md:136]: The content filled out in the code block does not need to be included `Signed-off-by` and `Co-authored-by`.

Here is an actual example：https://github.com/ti-community-infra/test-dev/pull/497

File: votes/0162-new-pr-requirement-and-commit-message-style.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
+# New Pull Request Requirement and Commit Message style
+
+Author: @zhangyangyu @Mini256
+
+## Motivation
+
+Currently for `tikv` and `pd` repositories, when a pull request is merging, the final commit message consists of an one-line subject, which is the pull request title, and body. The body rule is complex and messy. When a pull request has multiple commits, the bot uses the GitHub squashed commit message as the final commit message body. And due to the DCO requirement, every commit in the pull request needs to have a DCO signature. The signatures of each commit are also extracted into the final commit message body. A common final commit message now is like:
```
Comment (siddontang) [inline votes/0162-new-pr-requirement-and-commit-message-style.md:7]: s/is merging/is merged/

File: votes/0162-new-pr-requirement-and-commit-message-style.md (hunk 1 above)
Comment (sunxiaoguang) [reply votes/0162-new-pr-requirement-and-commit-message-style.md:136]: So we can remove signoff messages from code block in this example.

**Review by BusyJay — APPROVED** [review]
Amazing! I have been wanted the feature for quite a long time!

File: votes/0162-new-pr-requirement-and-commit-message-style.md (hunk 1 above)
Comment (zhangyangyu) [reply votes/0162-new-pr-requirement-and-commit-message-style.md:136]: This example is not the code block in the pull request body. It's the improved final commit message for the first sample in the Motivation section.

Comment (sunxiaoguang) [reply votes/0162-new-pr-requirement-and-commit-message-style.md:136]: Oops, I was reading original md and didn't render it in my mind😂

File: votes/0162-new-pr-requirement-and-commit-message-style.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+And if the pull request body contains the issue information `Issue Number: close #12341, close #12342` and the sample `commit-message` code block, the final commit message will be:
+
+```
+rsmeter: add pubsub datasink  (#11719)
+
+any multiple line commit messages that go into
+the final commit message body
+
+* fix something 1
```
Comment (nolouch) [inline votes/0162-new-pr-requirement-and-commit-message-style.md:148]: Does this part still contain some useless information like `"format code", "address comments", "add unit tests"`?

**Review by nolouch — COMMENTED** [review]
Great work!

File: votes/0162-new-pr-requirement-and-commit-message-style.md (hunk 3 above)
Comment (sunxiaoguang) [reply votes/0162-new-pr-requirement-and-commit-message-style.md:148]: I think what it does is that only messages quoted in code block are concatenated into final commit messages. 

Therefore you only see useful messages that are explicitly quoted in commit.

Comment (nolouch) [reply votes/0162-new-pr-requirement-and-commit-message-style.md:148]: Ah, get it. so the author or co-authored need to write `commit message` code block before merging.

Comment (Mini256) [reply votes/0162-new-pr-requirement-and-commit-message-style.md:148]: > Ah, get it. so the author or co-authored need to write commit message code block before merging.

The robot will fetch the commit list and take the commit authors (except PR author) as co-authors, automatically populate to the final commit message, so we don't need to manually fill in the code block.

**Comment by BusyJay** [comment]
Because author can also update the PR body, I think we should abort merging if the PR body, or the associated issue numbers and commit-message block more precisely, is changed after `/merge` is commented.

**Comment by Mini256** [comment]
> Because author can also update the PR body, I think we should abort merging if the PR body, or the associated issue numbers and commit-message block more precisely, is changed after /merge is commented.

Great suggestion, it will be implemented in the merge plugin.

**Review by nolouch — APPROVED** [review]
lgtm

**Review by sunxiaoguang — APPROVED** [review]

**Review by rleungx — APPROVED** [review]
LGTM

**Pull request merged**